server/src/main/java/io/druid/server/coordinator/cost/SegmentCostCache.java

```diff
@@ -0,0 +1,379 @@
+
+    public int size()
+    {
+      return buckets.size();
+    }
+
+    public SegmentCostCache build()
+    {
+      return new SegmentCostCache(
+          buckets
+              .entrySet()
+              .stream()
+              .map(entry -> entry.getValue().build())
+              .collect(Collectors.toCollection(ArrayList::new))
+      );
+    }
+
+    private Interval toBucketInterval(DataSegment segment)
+    {
+      long start = segment.getInterval().getStartMillis() - (segment.getInterval().getStartMillis() / BUCKET_INTERVAL);
+      return new Interval(start, start + BUCKET_INTERVAL);
+    }
+  }
+
+  static class Bucket
+  {
+    private final Interval interval;
+    private final ArrayList<DataSegment> sortedSegments;
+    private final double[] leftSum;
+    private final double[] rightSum;
+
+    public Bucket(Interval interval, ArrayList<DataSegment> sortedSegments, double[] leftSum, double[] rightSum)
+    {
+      this.interval = Preconditions.checkNotNull(interval, "interval");
+      this.sortedSegments = Preconditions.checkNotNull(sortedSegments, "sortedSegments");
+      this.leftSum = Preconditions.checkNotNull(leftSum, "leftSum");
+      this.rightSum = Preconditions.checkNotNull(rightSum, "rightSum");
+      Preconditions.checkArgument(sortedSegments.size() == leftSum.length && sortedSegments.size() == rightSum.length);
+      Preconditions.checkArgument(Ordering.from(SEGMENT_INTERVAL_COMPARATOR).isOrdered(sortedSegments));
+    }
+
+    public Interval getInterval()
+    {
+      return interval;
+    }
+
+    public boolean inCalculationInterval(DataSegment dataSegment)
+    {
+      return Math.abs(dataSegment.getInterval().getStartMillis() - interval.getStartMillis()) < LIFE_THRESHOLD
+             && Math.abs(dataSegment.getInterval().getEndMillis() - interval.getStartMillis()) < LIFE_THRESHOLD;
+    }
+
+    public double cost(DataSegment dataSegment)
+    {
+      double cost = 0.0;
+
+      double t0 = convertStart(dataSegment, interval);
+      double t1 = convertEnd(dataSegment, interval);
+
+      // avoid calculation for segments outside of LIFE_THRESHOLD
+      if (!inCalculationInterval(dataSegment)) {
+        return cost;
+      }
+
+      int index = Collections.binarySearch(sortedSegments, dataSegment, SEGMENT_INTERVAL_COMPARATOR);
+      index = (index >= 0) ? index : -index - 1;
+
+      // add to cost all left-overlapping segments
+      int leftIndex = index - 1;
+      while (leftIndex >= 0 && sortedSegments.get(leftIndex).getInterval().overlaps(dataSegment.getInterval())) {
+        double start = convertStart(sortedSegments.get(leftIndex), interval);
+        double end = convertEnd(sortedSegments.get(leftIndex), interval);
+        cost += CostBalancerStrategy.intervalCost(end - start, t0 - start, t1 - start);
+        --leftIndex;
+      }
+      // add left-non-overlapping segments
+      if (leftIndex >= 0) {
+        cost += leftSum[leftIndex] * (FastMath.exp(-t1) - FastMath.exp(-t0));
```

Comment:
FastMath is used in CostBalancerStrategy, I assume it has been done intentionally and there are some considerations behind. But I did not do any benchmarks myself, so can not be sure. Do you see any drawbacks in using FastMath?